sys/parsoid.js

```diff
@@ -455,6 +455,67 @@ class ParsoidService {
         return true;
     }
 
+    getHtml(hyper, req) {
+        if (req.query.sections) {
+            return this.getSections(hyper, req);
+        }
+        if (mwUtil.isNoCacheRequest(req)) {
+            // assume this is an update request
+            return this.getFormat('html', hyper, req);
+        }
+        return mwUtil.getSiteInfo(hyper, req).then((siteinfo) => {
+            if (!siteinfo.extensions['Flagged Revisions']) {
+                // the project does not have the FlaggedRevs extension
+                // enabled so proceed as normal
+                return this.getFormat('html', hyper, req);
+            }
+            const rp = req.params;
+            const queryReq = {
+                uri: new URI([rp.domain, 'sys', 'action', 'query']),
+                method: 'post',
+                body: {
+                    prop: 'flagged',
+                    formatversion: 2
+                }
+            };
+            if (rp.revision) {
+                queryReq.body.revids = rp.revision;
+            } else {
+                queryReq.body.titles = rp.title;
+            }
+            return P.props({
+                html: this.getFormat('html', hyper, req),
+                revinfo: hyper.post(queryReq)
+            }).then((res) => {
+                const html = res.html;
+                const retrev = mwUtil.parseETag(html.headers.etag).rev;
+                const revinfo = res.revinfo.body.items[0];
+                if (!revinfo.flagged || revinfo.flagged.stable_revid === retrev) {
+                    // the revision is not flagged, or the client is
+                    // fetching a stable revision, so it's safe to return it
+                    return html;
+                }
+                if (!rp.revision) {
+                    // the client is looking for the latest revision, so
+                    // redo the request with the stable revision
+                    req.params.revision = revinfo.flagged.stable_revid;
+                    return this.getFormat('html', hyper, req);
+                }
+                // the client supplied a revision, but that one is
+                // flagged, so bail out
+                // XXX: this is a problematic case, we need to find
+                // a solution for this
```

Comment:
This sounds like a flagged revision is a bad thing. While I'm sure my understanding of flagged revisions is not complete, I have a different understanding of flagged revisions. If a revision is flagged it's marked as good/reviewed, even when just DeferredChanges is enabled, AFAICS.

Example: http://deferred-changes.wmflabs.org/wiki/Special:ApiSandbox#action=query&format=json&prop=flagged&meta=&titles=Test_passive_deferred
returns `"stable_revid": 19`

http://deferred-changes.wmflabs.org/w/index.php?title=Test_passive_deferred&action=history
has two changes:
* rev 19: reviewed as good/flagged
* rev 20: unreviewed/not flagged

Reply:
Yeah, expected behavior for the explicit revision request should be to just return the content, as before. This ensures that these revisions can still be edited, among other things.

Reply:
Sounds like the real problem is that something else links to the latest revision instead of the non-revision-specific URL. That cannot be fixed here, and just returning the last revision when asked to do so will not make anything worse (crawlers getting the last revision instead of the stable one via RESTBase is already the status quo, right?).

Reply:
I was mostly looking at the comments. Once they get updated with something like @tgr's proposals then most parts start making more sense to me.